Assets/Shopify/UIToolkit/Components/WebImageCache.cs

```diff
@@ -0,0 +1,156 @@
+        /// <returns>The cached resource associated with the given url.</returns>
+        public CachedWebResource<Texture2D>? TextureResourceForURL(string url) {
+            var texture = TextureFromCacheForURL(url);
+            if (texture != null) {
+                PromoteURL(url);
+            }
+            return texture;
+        }
+
+        /// <summary>
+        /// Associates the given URL with the given texture in the cache.
+        /// </summary>
+        /// <param name="url">URL to key the cache with.</param>
+        /// <param name="textureResource">A CachedWebResource<Texture2D> instance to cache in memory.</param>
+        public void SetTextureResourceForURL(string url, CachedWebResource<Texture2D> textureResource) {
+            // If we already have it in the list, promote it. If not then we'll need to add it.
+            if (TextureFromCacheForURL(url) != null) {
+                PromoteURL(url);
+                RemoveURLFromCache(url);
+            } else {
+                _recentlyUsed.AddFirst(new LinkedListNode<string>(url));
+            }
+
+            var estimatedTextureSize = EstimateMemoryFootprintForTexture(textureResource.Data);
+            int nextMemoryFootprint = EstimatedMemorySize + estimatedTextureSize;
+
+            // Evict the least recently used textures from the cache until we can fit the new one in.
+            while (nextMemoryFootprint > MemorySizeLimit) {
+                var oldestNode = _recentlyUsed.Last;
+                var oldestUsedTexture = _urlTextureCache[oldestNode.Value].Data;
+                var oldestUsedTextureSize = EstimateMemoryFootprintForTexture(oldestUsedTexture);
+                _urlTextureCache.Remove(oldestNode.Value);
+                _recentlyUsed.RemoveLast();
+                nextMemoryFootprint -= oldestUsedTextureSize;
+            }
+
+            EstimatedMemorySize = nextMemoryFootprint;
+            _urlTextureCache[url] = textureResource;
+        }
+
+        /// <summary>
+        /// Removes the URL and associated Texture2D instance from the cache.
+        /// </summary>
+        /// <param name="url">URL key to remove from the cache.</param>
+        public void RemoveURL(string url) {
+            RemoveURLFromCache(url);
+
+            var node =_recentlyUsed.Find(url);
+            if (node != null) {
+                _recentlyUsed.Remove(node);
+            }
+        }
+
+        /// <summary>
+        /// Clears the cache. 
+        /// </summary>
+        public void Clear() {
+            _urlTextureCache.Clear();
+            _recentlyUsed.Clear();
+            EstimatedMemorySize = 0;
+        }
+
+        private int EstimateMemoryFootprintForTexture(Texture2D texture) {
+            // Since we're only dealing with JPEG (RGB24) or PNG (ARGB32) formats, we can assume 4 bytes per.
+            // We can also assume these images will only have a single mipmap level.
+            return texture.width * texture.height * 4;
```

Comment:
Could be made more accurate by checking https://docs.unity3d.com/ScriptReference/Texture2D-format.html for a few common formats, and then fallback on assuming 4.